Sources/Core/CUDA/CUDAPCISPHSolver2.cpp

```diff
@@ -0,0 +1,162 @@
+// personal capacity and are not conveying any rights to any intellectual
+// property of any third parties.
+
+#if CUBBYFLOW_USE_CUDA
+
+#include <Core/CUDA/CUDAPCISPHSolver2.hpp>
+#include <Core/Particle/SPHKernels.hpp>
+#include <Core/PointGenerator/TrianglePointGenerator.hpp>
+
+using namespace CubbyFlow;
+
+// Heuristically chosen value.
+constexpr float DEFAULT_TIME_STEP_LIMIT_SCALE = 5.0f;
+
+CUDAPCISPHSolver2::CUDAPCISPHSolver2()
+{
+    SetTimeStepLimitScale(DEFAULT_TIME_STEP_LIMIT_SCALE);
+}
+
+CUDAPCISPHSolver2::CUDAPCISPHSolver2(float targetDensity, float targetSpacing,
+                                     float relativeKernelRadius)
+    : CUDASPHSolverBase2{}
+{
+    const auto sph = SPHSystemData();
+    sph->SetTargetDensity(targetDensity);
+    sph->SetTargetSpacing(targetSpacing);
+    sph->SetRelativeKernelRadius(relativeKernelRadius);
+
+    m_tempPositionsIdx = sph->AddVectorData();
+    m_tempVelocitiesIdx = sph->AddVectorData();
+    m_tempDensitiesIdx = sph->AddFloatData();
+    m_pressureForcesIdx = sph->AddVectorData();
+    m_densityErrorsIdx = sph->AddFloatData();
+
+    SetTimeStepLimitScale(DEFAULT_TIME_STEP_LIMIT_SCALE);
+}
+
+float CUDAPCISPHSolver2::MaxDensityErrorRatio() const
+{
+    return m_maxDensityErrorRatio;
+}
+
+void CUDAPCISPHSolver2::SetMaxDensityErrorRatio(float ratio)
+{
+    m_maxDensityErrorRatio = std::max(ratio, 0.0f);
+}
+
+unsigned int CUDAPCISPHSolver2::MaxNumberOfIterations() const
+{
+    return m_maxNumberOfIterations;
+}
+
+void CUDAPCISPHSolver2::SetMaxNumberOfIterations(unsigned int n)
+{
+    m_maxNumberOfIterations = n;
+}
+
+CUDAArrayView1<float2> CUDAPCISPHSolver2::TempPositions()
+{
+    return SPHSystemData()->VectorDataAt(m_tempPositionsIdx);
+}
+
+CUDAArrayView1<float2> CUDAPCISPHSolver2::TempVelocities()
+{
+    return SPHSystemData()->VectorDataAt(m_tempVelocitiesIdx);
+}
+
+CUDAArrayView1<float> CUDAPCISPHSolver2::TempDensities()
+{
+    return SPHSystemData()->FloatDataAt(m_tempDensitiesIdx);
+}
+
+CUDAArrayView1<float2> CUDAPCISPHSolver2::PressureForces()
+{
+    return SPHSystemData()->VectorDataAt(m_pressureForcesIdx);
+}
+
+CUDAArrayView1<float> CUDAPCISPHSolver2::DensityErrors()
+{
+    return SPHSystemData()->FloatDataAt(m_densityErrorsIdx);
+}
+
+float CUDAPCISPHSolver2::ComputeDelta(float timeStepInSeconds)
+{
+    const auto particles = SPHSystemData();
+    const auto kernelRadius = static_cast<double>(particles->KernelRadius());
+
+    Array1<Vector2D> points;
+    const TrianglePointGenerator pointsGenerator;
+    const Vector2D origin;
+    BoundingBox2D sampleBound(origin, origin);
+    sampleBound.Expand(1.5 * kernelRadius);
+
+    pointsGenerator.Generate(
+        sampleBound, static_cast<double>(particles->TargetSpacing()), &points);
+
+    const SPHSpikyKernel2 kernel(kernelRadius);
+
+    double denom = 0;
+    Vector2D denom1;
+    double denom2 = 0;
+
+    for (size_t i = 0; i < points.Length(); ++i)
+    {
+        const Vector2D& point = points[i];
```

Comment:
```suggestion
    for (const auto& point : points)
    {
```
I found that index `i` used only for indexing `Array1<Vector2D> points`.
How about replace it with a range-based for loop 😉
Also same things in [`CUDAPCISPHSolver3.cpp`](https://github.com/CubbyFlow/CubbyFlow/pull/117/files#diff-065eebb8c7d8d03b9719a77fba44df304dcaa9b767dee4d96b0e791811630dc7) can be done.

Reply:
Resolves in 81dca5e. 🕊️

Reply:
Cool!